#42 Expect messages from either source address 0x00 or 0x01

gumbelmj opened the pull request:
This changes the logic to not rely on just source address 0 for information.  We will use the highest value for the ignition cycles or OBD counts.  We will use the first monitors packet or ratios packet as the initial time.  We will re-try the requests for DM5, DM20, DM21 and DM26 until we get a response from either source address 0 or source address 1.  This will also fix #45, fix #47 and fix #42

gumbelmj committed:
#42 Expect messages from either source address 0x00 or 0x01

gumbelmj commented:
fix #42

tgolub1 commented:
I am unsure of the etiquette re comments to this project, so if I am out of bounds posting here, please let me know.  With that said - the checkin comment "...until we get a response from either source address 0 or source address 1" leaves additional concerns.  Something along the lines of "until we get responses from SA 1 and SA 61" would be better represent the expectations for Detroit powertrain.

An over-all generic approach could also be to:
1.  Issue a GLOB RQST for DM5 and generate a list of SAs which respond with control byte 0x14
2.  Issue DS RQSTs to all those SAs for the DMx in question

gumbelmj commented:
@tgolub1 with the original design, we were addresss agnostic.  After our first round of testing, we discovered an issue with the engine not always responding to requests.  Thus we added logic to retry requests if the engine (SA 0) did not respond per the suggestion of our sponsors.  I changed the logic to 'always' expect a messages from SA0 or SA1.  (This only applies to DM5, DM20, DM21 and DM26.)

Earlier in the design, I recommended making all (or nearly all) requests DS to the OBD Modules.  I was instructed to not do that and make the requests global.  Doing DS requests would also speed up the process as we currently have to wait a preset time to any modules that may or may not respond.

We have another issue that hasn't been logged yet to change the order of operations to do as you have recommended (use the DM5 to make DS request to OBD Modules). If we implement that as suggested, then the logic to rely on SA0 or SA1 would be removed.

However at this time we need to talk with the sponsors to determine if such a change is really in scope or out of scope, given that we had specific instructions for our implementation.

At this time, if you'd like you can clone this project and with a JDK and Ant installed on a Windows machine try out the latest code to determine if it will work with DDC engines.

tgolub1 commented:
@gumbelmj Thanks for the clarification!  I would personally be in favor of your original DS RQST approach for DM messages.  Especially given that a lack of a response to a GLOB RQST really *cannot* be taken at a face value.  Please see the attached J1939-71 specification fragment:

![image](https://user-images.githubusercontent.com/31372094/29848178-103dd632-8ced-11e7-8602-5524bd19d6ab.png)

As can be seen:
-  Devices will simply drop a GLOB RQST for non-supported DMs
-  Devices will / could simply drop a GLOB RQST for a supported DM if a TP-BAM session is already active (due to unrelated periodic Tx)
-  Devices will / could simply drop GLOB RQST for some specific DMs (eg. DM24, DM25, DM30, etc?) to conserve network bandwidth.  For at least these, diag tool must issue a DS RQST

gumbelmj commented:
@tgolub1 Thanks for the info.  We will continue to push for this. In the meantime, I logged the issue as #53 so we don't lose track of it.

tgolub1 commented:
@gumbelmj Using ant I can generate a collection of class files in the build directory.  Is there a specific manifest that should be used to combine these into a jar file?  (What is the actual entry point?)

gumbelmj commented:
@tgolub1 running the ant task "dist" should produce the executable and the installer.  They will be in the "dist" directory.  The [ReadMe](https://github.com/battjt/iumpr#contribute) should have enough information to get started.  If it doesn't please let me know.
